editor/store/selectors.js

```diff
@@ -564,6 +565,62 @@ export const getBlocks = createSelector(
 	]
 );
 
+/**
+ * Given a block, returns an array containing all clientIds
+ * of referenced blocks (if it's a shared block)
+ * and innerBlocks.
+ *
+ * @param {Object} state Global application state.
+ * @param {Object} block The block whose ids we want to unfold.
+ *
+ * @return {Array} ids of referenced and inner blocks.
+ */
+const unfoldClientIds = ( state, block ) => {
+	if ( ! block ) {
+		return [ null ];
+	}
+
+	const getClientIdsFromSharedBlock = ( globalState, sharedBlock ) => {
+		if ( sharedBlock.name === 'core/block' ) {
+			const clientId = get( getSharedBlock( globalState, sharedBlock.attributes.ref ), [ 'clientId' ], null );
+			return [
+				clientId,
+				...unfoldClientIds( globalState, ...getBlocksByClientId( globalState, clientId ) ),
+			];
+		}
+		return [ null ];
```

Comment:
If we just return [] here we avoid the need for the filter by id right?

Reply:
I don't think so: in https://github.com/WordPress/gutenberg/pull/8075/files/90ff8269cf5999dd605cb71a51251846e3354efc#diff-4e1dd4776f75eb7a66b28280a56d8e0aR594 and https://github.com/WordPress/gutenberg/pull/8075/files/90ff8269cf5999dd605cb71a51251846e3354efc#diff-4e1dd4776f75eb7a66b28280a56d8e0aR585 we return `null` if the shared or inner block doesn't have a `clientId`. That's certainly unlikely, but I'd rather be cautious.

Reply:
Thank you for clarigying, I missed that the "block" it self may be null because of the return of a selector.